nx_cugraph/algorithms/link_prediction.py

```diff
@@ -0,0 +1,99 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import cupy as cp
+import networkx as nx
+import pylibcugraph as plc
+
+from nx_cugraph.convert import _to_undirected_graph
+from nx_cugraph.utils import networkx_algorithm, not_implemented_for
+
+__all__ = [
+    "jaccard_coefficient",
+]
+
+
+@not_implemented_for("directed")
+@not_implemented_for("multigraph")
+@networkx_algorithm(version_added="25.02", _plc="jaccard_coefficients")
+def jaccard_coefficient(G, ebunch=None):
+    if ebunch is None:
+        # FIXME: is there a more efficient way to do this (on GPU or
+        # otherwise)?
+        ebunch = list(nx.non_edges(G))
+        if not ebunch:
+            return iter([])
```

Comment:
Here's an alternative:
```suggestion
    G = _to_undirected_graph(G)
    if ebunch is None:
        A = cp.tri(G._N, G._N, dtype=bool)
        A[G.src_indices, G.dst_indices] = True
        src_indices, dst_indices = cp.nonzero(~A)
        if src_indices.size == 0:
            return iter([])
        src_indices = src_indices.astype(index_dtype)
        dst_indices = dst_indices.astype(index_dtype)
```
there are other variations and alternatives; for example, you could use a mask:
```python
        mask = G.src_indices < G.dst_indices
        A[G.src_indices[mask], G.dst_indices[mask]] = True
```
Observe that these go straight to indices, so there is no need to call e.g. `G._list_to_nodearray(src_indices)` to do remapping from keys to indices.

Also, if possible, I would advise against using the graph object (such as calling `nx.non_edges(G)`) before converting the graph via e.g. `G = _to_undirected_graph(G)`. I think it's best to make fewer assumptions about the input.

Reply:
ok thanks, I'm going to benchmark and apply one of these suggestions (if they're faster, which I'm assuming they are).

I'm a bit confused by this statement though:
> I would advise against using the graph object (such as calling nx.non_edges(G)) before converting the graph via e.g. G = _to_undirected_graph(G). I think it's best to make fewer assumptions about the input.

If I need to call a NX function like `nx.non_edges()`, why would I first convert the input to a `CudaGraph` using `_to_undirected_graph()`?  Shouldn't we be assuming the input is a compatible NX `Graph` type, or are we allowing `CudaGraph` objects to be passed as well?  I suspect I'm missing something. (side note: this is ideal info for a docstring, but I believe the decorator bubbles them up to the corresponding nx function docstring, which I wouldn't want here.)